solr/core/src/java/org/apache/solr/handler/admin/HealthCheckHandler.java

```diff
@@ -88,15 +95,42 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
       return;
     }
 
-    // Set status to true if this node is in live_nodes
-    if (clusterState.getLiveNodes().contains(cores.getZkController().getNodeName())) {
-      rsp.add(STATUS, OK);
-    } else {
+    // Fail if not in live_nodes
+    if (!clusterState.getLiveNodes().contains(cores.getZkController().getNodeName())) {
       rsp.add(STATUS, FAILURE);
       rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE, "Host Unavailable: Not in live nodes as per zk"));
+      return;
     }
 
-    rsp.setHttpCaching(false);
+    // Optionally require that all cores on this node are active if param 'failWhenRecovering=true'
+    if (req.getParams().getBool(PARAM_REQUIRE_HEALTHY_CORES, false)) {
+      List<String> unhealthyCores = findUnhealthyCores(clusterState, cores.getNodeConfig().getNodeName());
+      if (unhealthyCores.size() > 0) {
+          rsp.add(STATUS, FAILURE);
+          rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE,
+                  "Replica(s) " + unhealthyCores + " are currently initializing or recovering"));
+          return;
+      }
+      rsp.add("MESSAGE", "All cores are healthy");
+    }
+
+    // All lights green, report healthy
+    rsp.add(STATUS, OK);
+  }
+
+  /**
+   * Find replicas DOWN or RECOVERING
+   * @param clusterState clusterstate from ZK
+   * @param nodeName this node name
+   * @return list of core names that are either DOWN ore RECOVERING on 'nodeName'
+   */
+  static List<String> findUnhealthyCores(ClusterState clusterState, String nodeName) {
+    return clusterState.getCollectionsMap().values().stream()
```

Comment:
Not sure about performance here if you have hundreds of collections and thousands of replicas in a large cluster.
The alternative is to instead iterate cores on current node, and consult with clusterState their overall state.

Reply:
> maybe we want to return false if there are any replicas from inactive slices on the node

Inactive shards are not searched, so we should not care about those. A shard split will not clean up the old shard, but instead mark it inactive, until user manually deletes those shards, or the Autoscaling framework rules go reap them. That is why I chose to check active shards only. We should be ok if the active shard(s) only are up and active, then k8s can go restart the next node.

If a shard split is currently running (could be long running), on a node being restarted, the split would be aborted but when the node comes up again I believe the overseer might try again??

Reply:
> Also if the clusterState thinks that cores live on this node, but the core directories do not exist, then I think that this handler should respond not healthy.

Yes, we can add an extra check that for each replica in clusterstate for node, we check that it exists locally?

Reply:
> > Also if the clusterState thinks that cores live on this node, but the core directories do not exist, then I think that this handler should respond not healthy.
> 
> Yes, we can add an extra check that for each replica in clusterstate for node, we check that it exists locally?

I think that the logic as it stands now should work, because the cluster state will report the replica as "DOWN" (From my experience, but we can also add tests around this). The comment was meant to validate the current approach over the other one you mentioned:

> The alternative is to instead iterate cores on current node, and consult with clusterState their overall state.

Reply:
> If a shard split is currently running (could be long running), on a node being restarted, the split would be aborted but when the node comes up again I believe the overseer might try again??

And I understand your decision for the active shards better now. As long as the overseer thing is true, then we should be fine. And if in the future if we need to, we can add another parameter to fail on inactive slices as well.